Feature/refactor BaseAdaptor

markmcdowall opened the pull request:
## Requirements

- To refactor the BaseAdaptor inline with perlcritic and podchecker

## Description

Updated the BaseAdaptor. This includes:
- Documentation for all functions
- Use of prepare_cached so that the prepare statements are cached once prepared
- Refactored the upload_xref_object_graphs function so that it is less complex. This was done by using either methods that were already available to reduce code duplication or extracting SQL calls into their own functions so that they are accessible from elsewhere.
- SQL tidying
- Code tidying
- Converted the errors so that they confess rather than croak so that the issues they raise are not missed. If there are errors then it should be up to the parser to handle them.

## Use case

This should hopefully reduce the code complexity and improve long term development

## Benefits

Reduced technical debt

## Possible Drawbacks

There will need to be changes in parsers so that they are able to handle errors that are thrown when they used to be ignored

## Testing

I have run the code tests that are available for the parsers.

The ArrayExpress parser test now fails and will require minor changes

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -899,49 +893,74 @@ sub get_direct_xref {
     croak( $self->dbi->errstr() );
   if ( my @row = $get_xref_sth->fetchrow_array() ) {
+
+    # Calling finish() as we only requires the first row only
```
Typo

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -950,43 +969,58 @@ sub get_object_xref {
+=head2 add_xref
+  Arg [1]    : xref
+  Description: Create an xref
+               If it already exists it return that xrefs xref_id else creates it
```
Typo, missing apostrophe, s/else/otherwise/ for a more conventional idiom

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1001,20 +1035,26 @@ sub add_xref {
                           $description, $source_id, $species_id,
                           $info_type,   $info_text ) or
-    croak("$acc\t$label\t\t$source_id\t$species_id\n");
+    confess "$acc\t$label\t\t$source_id\t$species_id\n";
```
I wonder if we could get the DBI error into this message?

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1450,80 +1711,107 @@ GDH
+
+=head2 set_release
+  Arg [1]    : source ID
+  Arg [2]    : ensembl release
```
Not an Ensembl release, but the relevant release number for the Xref source

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1268,20 +1384,79 @@ ADX
+=head2 add_multiple_dependent_xrefs
+  Arg [1]    : xref ID
+  Arg [2]    : xref
+  Description: Add dependent xrefs to the xref table aong with dependent xref mappings
```
Typo

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1307,51 +1493,95 @@ sub add_to_syn {
   else {
-    carp( "Could not find acc $acc in " .
-          "xref table source = $source_id of species $species_id\n" );
+    confedd( "Could not find acc $acc in " .
```
Nope

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1307,51 +1493,95 @@ sub add_to_syn {
+
+=head2 add_multiple_synonyms
+  Arg [1]    : xref ID
+  Arg [2]    : synonyms
```
Listref of synonyms would be more informative

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1450,80 +1711,107 @@ GDH
+  Arg [1]    : source name
+  Description: Create a has that uses the accession and labels for keys and an
+               array of the synonyms as the values
+  Return type: hash
```
Not a hash

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1450,80 +1711,107 @@ GDH
+
+=head2 get_ext_synonyms
+  Arg [1]    : source name
+  Description: Create a has that uses the accession and labels for keys and an
```
Not a has.

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1554,77 +1842,227 @@ GES
 
-  # Store max id for
+=head2 parsing_finished_store_data
+  Description: Store data needed to beable to revert to same stage as after parsing
```
What is a beable?

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1554,77 +1842,227 @@ GES
+
+=head2 get_meta_value
+  Arg [1]    : key
+  Description: Get the last value from the meta data table that matches like the
```
As a matter of taste, I prefer a plain English description that cuts to the chase, rather than anglicised pseudocode.

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -1554,77 +1842,227 @@ GES
+=head2 _add_pair
+  Arg [1]    : source ID
+  Arg [2]    : accession
+  Arg [3]    : pair
```
Pair of what?

markmcdowall replied:
This is the RefSeq pair, now described in the line below

markmcdowall commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -403,277 +480,124 @@ sub get_valid_codes {
+    $self->add_multiple_direct_xrefs( $xref );
+
+    #############################################################################
+# create entry in primary_xref table with sequence; if this is a "cumulative"
```
This should be fixed as part of the #11 , but the change can be made here as well.

markmcdowall commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -65,43 +83,68 @@ sub new {
+=head2 new
+  Description: Getter/Setter for the dbc object
+  Return type: db_connection
+  Exceptions : none
```
All `Exceptions : none` will get removed

markmcdowall commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -117,107 +160,110 @@ sub get_filehandle {
   my $big_name = uc $source_name;
-  my $sql =
-"SELECT source_id FROM source WHERE upper(name) LIKE '%${big_name}%'";
+  my $sql = 'SELECT source_id FROM source WHERE UPPER(name) LIKE ?';
```
Maybe, but provides clarity between columns and keywords

markmcdowall commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -314,7 +380,7 @@ GDS
     $t_sql =~ s/TYPE/$type/xsm;
 
-    $sth[ $i++ ] = $self->dbi->prepare($t_sql);
+    $sth[ $i++ ] = $self->dbi->prepare_cached($t_sql);
```
I'll recode this as I have done with the other examples where there are multiple on the fly generated SQL

markmcdowall commented:
@nerdstrike, @magaliruffier, @premanand17 and @avullo 

I have frozen development on this branch for the unit testing. This will now be developed on the branch feature/unittests_BaseAdaptor so that these changes can be merged in without introducing new issues or delaying other development.

Since @nerdstrike reviewed the code I have addressed the changes suggested and have also removed duplicate INSERT SQL code from direct and dependent xref code functions.

markmcdowall commented:
With the merging of the feature/unittests_BaseAdaptor branch this pull request now includes better testing across the BaseAdaptor. There were minor changes to the BaseAdaptor that were highlighted by the creation of the tests.

premanand17 commented on modules/t/baseadaptor.t:
```diff
@@ -0,0 +1,571 @@
+
+ok(defined $source->source_id, 'Was the source created in the DB?');
+
+my $xref = $source->create_related('xrefs', {
```
why not pass $fake_xref here as argument instead of repeating

premanand17 commented on modules/t/baseadaptor.t:
```diff
@@ -0,0 +1,571 @@
+is( $xref_dba->get_source_name_for_source_id($source->source_id), 'RefSeq', 'get_source_name_for_source_id' );
+
+
+# get_valid_xrefs_for_dependencies - Tested later once dependecny xrefs are added
```
remove

premanand17 commented on modules/t/baseadaptor.t:
```diff
@@ -0,0 +1,571 @@
+# get_valid_xrefs_for_dependencies - Tested later once dependecny xrefs are added
+
+
+# get_valid_xrefs_for_direct_xrefs - Tested later once direct xrefs are added
```
remove

premanand17 commented on modules/t/baseadaptor.t:
```diff
@@ -0,0 +1,571 @@
+);
+
+
+# get_xref - Already tested for other loading functions
```
remove

premanand17 commented on modules/t/baseadaptor.t:
```diff
@@ -0,0 +1,571 @@
+# get_xref - Already tested for other loading functions
+
+
+# get_object_xref - Testing later once values added
```
remove

premanand17 commented:
As the changes requested by Keiron, has also been addressed, am going ahead to merge BaseAdaptor.  We need this in master to proceed with further mapping updates.

nerdstrike commented on modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm:
```diff
@@ -46,6 +50,7 @@ use Carp;
 use Bio::EnsEMBL::Utils::Exception;
 use Bio::EnsEMBL::Xref::FetchFiles;
 use Getopt::Long;
```
Sorry no, this module is only used for command line interfaces. It has no business here.
